AxelarHandler/src/GoFastHandler.sol

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import {IERC20} from "lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "lib/openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {ISwapRouter02} from "./interfaces/ISwapRouter02.sol";
+import {IFastTransferGateway} from "./interfaces/IFastTransferGateway.sol";
+
+contract GoFastHandler {
+    using SafeERC20 for IERC20;
+
+    ISwapRouter02 public swapRouter;
+    IFastTransferGateway public fastTransferGateway;
+
+    constructor(address _swapRouter, address _fastTransferGateway) {
+        swapRouter = ISwapRouter02(_swapRouter);
+        fastTransferGateway = IFastTransferGateway(_fastTransferGateway);
+    }
+
+    function swapAndSubmitOrder(
+        address tokenIn,
+        uint256 swapAmountIn,
+        bytes memory swapCalldata,
+        uint256 feeAmount,
+        bytes32 sender,
+        bytes32 recipient,
+        uint32 destinationDomain,
+        uint64 timeoutTimestamp,
+        bytes calldata destinationCalldata
+    ) public payable returns (bytes32) {
+        require(feeAmount != 0, "fast transfer fee cannot be zero");
+
+        uint256 swapAmountOut = _swap(tokenIn, swapAmountIn, swapCalldata);
+
+        require(swapAmountOut >= feeAmount, "amount received from swap is less than fast transfer fee");
+
+        // this is the amount that the recipient will receive on the destination chain
+        uint256 swapAmountOutAfterFee = swapAmountOut - feeAmount;
+
+        return fastTransferGateway.submitOrder(
+            sender,
+            recipient,
+            swapAmountOut,
+            swapAmountOutAfterFee,
+            destinationDomain,
+            timeoutTimestamp,
+            destinationCalldata
+        );
+    }
+
+    function _swap(address tokenIn, uint256 amountIn, bytes memory swapCalldata) internal returns (uint256 amountOut) {
+        address tokenOut = fastTransferGateway.token();
+
+        uint256 tokenOutBalanceBefore = IERC20(tokenOut).balanceOf(address(this));
+
+        if (tokenIn != address(0)) {
```

Comment:
So the zero address is reserved for native eth inputs? I may be out of date, but I thought there was some string people use by convention, like something like 0xEEE...

Reply:
Both are used by convention, which makes things not really a convention, but our other contracts already use `address(0)`.